src/antelope/chains/Token.ts

```diff
@@ -0,0 +1,125 @@
+    readonly isNative: boolean;    // True if the token is a native blockchain token
+    private _price: number;        // Token price in USD
+
+    constructor(sourceInfo: TokenSourceInfo) {
+        this.id = `${sourceInfo.symbol}-${sourceInfo.contract}-${sourceInfo.chainId}`;
+        this.symbol = sourceInfo.symbol;
+        this.contract = sourceInfo.contract ?? sourceInfo.address ?? '';
+        this.chainId = sourceInfo.chainId;
+        this.name = sourceInfo.name;
+        this.decimals = sourceInfo.decimals ?? sourceInfo.precision ?? 0;
+        this.isSystem = sourceInfo.isSystem;
+        this.isNative = sourceInfo.isNative;
+        this.logo = sourceInfo.logo ?? sourceInfo.logoURI;
+        this._price = 0;
+    }
+
+    // Returns the URI for the token logo
+    get logoURI(): string | undefined {
+        return this.logo;
+    }
+
+    // Returns the token price
+    get price(): number {
+        return this._price;
+    }
+
+    // Updates the token price
+    updatePrice(price: number) {
+        this._price = price;
+    }
+}
+
+// A class to represent an EVM blockchain token
+export class EvmToken extends Token {
+    private _metadata?: string;    // Token contract metadata
+    readonly type?: EvmTokenType;  // Token type (ERC20, ERC721, etc.)
+
+    constructor(sourceInfo: TokenSourceInfo) {
+        super(sourceInfo);
+        this._metadata = sourceInfo.metadata;
+    }
+
+    // Returns the token address (same as contract address)
+    get address(): string {
+        return this.contract;
+    }
+
+    // Returns the token metadata
+    get metadata(): string | undefined {
+        return this._metadata;
+    }
+
+    // Returns if the token has metadata
+    get hasMetadata(): boolean {
+        return !!this._metadata;
+    }
+
+    // Sets the token metadata
+    setMetadata(metadata: string) {
+        this._metadata = metadata;
+    }
+}
+
+// A class to represent a native blockchain token
+export class NativeToken extends Token {
+    readonly precision: number;
+
+    constructor(sourceInfo: TokenSourceInfo) {
+        super(sourceInfo);
+        this.precision = sourceInfo.precision || 0;
+    }
+}
+
+// A class to represent the balance of a token
+export class TokenBalance {
+    readonly token: Token;
+    readonly balance: string;
+    readonly balanceBn: ethers.BigNumber;
+
+    constructor(token: Token, balanceBn: ethers.BigNumber) {
+        this.token = token;
+        this.balanceBn = balanceBn;
+        this.balance = balanceBn.toString(); // Convert balanceBn to a string representation
+    }
+
+    // Returns the fiat balance based on the current token price and balance
+    get fiatBalance(): ethers.BigNumber {
+        const price = ethers.utils.parseUnits(this.token.price.toString(), 18);
+        const fiatDouble = this.balanceBn.mul(price);
+        const fiat = fiatDouble.div(ethers.utils.parseUnits('1', 18));
+        return fiat;
+    }
```

Comment:
i actually don't understand how this conversion works :flushed: i could only do it using my `convertCurrency` method from the CurrencyInput PR. 

also i guess `fiatBalance` as a bignumber should have 2 decimals right? so a value of 1 USD would be equivalent to `BigNumber.from(100)`?

i think `fiatBalance` and `_price` should be the same type, right? because the `_price` is a fiat amount, and so is `fiatBalance`... they should probably both be either `number` or `BigNumber`. on one hand i like the idea of having all currency amounts being BigNumber because then there is no issue comparing and converting. but on the other hand it is sort of confusing to think of fiat as `BigNumber`. If you had to choose one type for both `_price` and `fiatBalance` what would you choose? i vote `BigNumber` but i would be okay with `number` i guess